src/zmalloc.h

```diff
@@ -136,6 +136,28 @@ size_t zmalloc_usable_size(void *ptr);
 #define zmalloc_usable_size(p) zmalloc_size(p)
 #endif
 
+void __attribute__((alloc_size(2))) __attribute__((noinline)) *extend_to_usable(void *ptr, size_t size);
+
+static inline size_t zmalloc_usable_size_safe(void **ptr) {
+    size_t sz = zmalloc_usable_size(*ptr);
+    *ptr = extend_to_usable(*ptr, sz);
+    return sz;
+}
+
+static inline void *zmalloc_safe(size_t size) {
+    size_t usable_size;
+    void *ptr = zmalloc_usable(size, &usable_size);
+    ptr = extend_to_usable(ptr, usable_size);
+    return ptr;
+}
+
+static inline void *zrealloc_safe(void *ptr, size_t size) {
+    size_t usable_size;
+    ptr = zrealloc_usable(ptr, size, &usable_size);
+    ptr = extend_to_usable(ptr, usable_size);
+    return ptr;
+}
```

Comment:
we already have too many allocation functions, and i'd like to avoid adding the `safe` option.
what i'd suggest to do instead is:
1. rename the current `z[try]malloc_usable` family to `z[try]malloc_internal` and don't expose them to users outside zmalloc.c
2. expose a new set of `zmalloc_usable` family that use `zmalloc_internal` and `extend_to_usable` implicitly.
3. go over all the users of `zmalloc_usable_size` and convert them to use the `zmalloc_usable` family if possible (as i think you did).
4. drop `zmalloc_usable_size_safe`, i don't like to pass a reference to a pointer, don't know if the caller kept the original somewhere, i'd rather have the `malloc` or `realloc` approach that returns it.